PS-3032 Put together a list view

joneshf-cn opened the pull request:
## What it does

This is the first go at having a list view. It follows the ideas laid out in our [spec][], but only takes a slice of the whole picture.

## Review asks

We're only looking to solve PS-3032 with this one. There's still a bunch of work for PS-3023, but is there anything we missed for PS-3032?

[spec]: https://www.notion.so/citizennet/List-View-Config-9a5c51094f9343ecb4a1615f869c29a5

joneshf-cn committed:
PS-3032 Start implementing list view

We start by defining the structure for list views.
This is a pared down implementation of what's in our specification:
https://www.notion.so/citizennet/List-View-Config-9a5c51094f9343ecb4a1615f869c29a5#5c95ed435d394f619a801c0e037e6fe2.
We're only concerning ourselves with what it will take to display
any arbitrary field.

We take a slightly different approach to how we do serialization.
In the past,
we've had the serialization layer and the business logic tightly coupled.
This works fine when we don't have any real dependents.
It doesn't work very well once we have dependents.
Any changes we make to the business logic are directly reflected in
the serialization.

This caused extra work for us most recently when we renamed some fields.
It meant that OMS had to wipe out a bunch of data in its data store.
Neverminding the fact that OMS should not be storing data in JSON blobs,
we still shouldn't be hostile to someone consuming our package.

To try and mitigate these problems in the future,
we start with a separation right now for list views.
Hopefully, when the code evolves,
it won't evolve in a way that forces us to constantly update dependents
for things they aren't really concerned about.

joneshf-cn committed:
PS-3032 Add a halogen interpretation

We want to be able to render these views.
However, we don't want our logic to be so coupled to halogen either.
That tends to bring quite a bit of incidental complexity base on what
halogen requires.
It's already requirinb quite a lot of stuff from us now,
and we're barely using it.

To mitigate the explosion of halogen, we make another seam for rendering.
The hope is that we don't have to constantly alter the core logic
whenever we need to alter the rendering.

We also add an example to our running examples that we've got.
There's not much to see, but this is where we are.

davezuch commented on src/Lynx/List.purs:
```diff
@@ -0,0 +1,216 @@
+    }
+
+eval :: forall f. Query ~> f
+eval query = eval query
```
How does this work?

joneshf-cn replied:
`Query _` is uninhabited. If someone were to construct a `Query _` and then passed it to `eval`, they've broken PureScript. There's no other implementation here than calling `eval` with the `Query _` they created.

joneshf-cn replied:
Let's just make `Query _` a synonym for `Const Void _`. Hopefully that requires less explanation in this interim stage before we get more behavior.

davezuch commented on src/Lynx/List.purs:
```diff
@@ -0,0 +1,216 @@
+fromArgonaut = Data.Codec.Argonaut.prismaticCodec from to Data.Codec.Argonaut.json
+  where
+  from :: Data.Argonaut.Json -> Maybe a
+  from json = case Data.Argonaut.decodeJson json of
```
```purescript
from = Data.Either.hush <<< Data.Argonaut.decodeJson
```
?

davezuch replied:
Or do we prefer the explicit casing?

joneshf-cn replied:
I prefer casing, but I'm down to switch.

davezuch replied:
Ah it's no big deal!

davezuch commented on src/Lynx/List.purs:
```diff
@@ -0,0 +1,216 @@
+-- | becoming out of sync.
+-- | It's a _mostly_ caught by the test suite,
+-- | but we can also allow the de/encoders to be correct by construction.
+jsonView :: Data.Codec.Argonaut.JsonCodec JSONView
```
This is really cool, I didn't know it was something Argonaut supported

davezuch commented on src/Lynx/List.purs:
```diff
@@ -0,0 +1,216 @@
+  column column' =
+    Ocelot.Block.Table.header_
+      [ Halogen.HTML.text (Lynx.Expr.print column'.name)
+      ]
```
I would think PS-3032 also involves rendering rows for a collection of `Array (Page a)` here, no?

joneshf-cn replied:
Makes sense, where would this live? Is it part of the list `View` or is it part of the `RenderView`?

davezuch replied:
So if I understand your setup, `View` and `RenderView` are analogous to `View Expr` and `View ExprType`, if we were to do things more similar to `Lynx.purs`, yeah? In that case we just need to add `value :: ExprType` (or `String`) to `RenderView`, and add an eva that goes from `View -> RenderView`.

joneshf-cn replied:
`View` and `RenderView` would be more like `Page Expr` and
```PureScript
type State
  = { activeTab :: String
    , evaled :: Either EvalError (Page ExprType)
    , expr :: Page Expr
    , fragment :: Fragment
    , idGenerator :: Aff String
    , values :: Map Key ExprType
    }
```
respectively. In any case, sounds like it belongs in the `RenderView`.

davezuch reviewed (commented):
This looks good so far, but I think part of PS-3032 is still missing, as I mentioned in a comment below.

joneshf-cn committed:
PS-3032 Inline casing

Per review request.

joneshf-cn committed:
PS-3032 Use `Const Void _` insted of defining a type

They're equivalent,
but hopefully this way we don't require an explanation.

joneshf-cn committed:
PS-3032 Provide a way to lookup a `Field _`

We're going to need to grab the `Field _` out of a `Page _`.
That `Field _` might not exist, but we can try it anyway.

Ideally, we'd have a function where you pass it a `Key`
and you get back `Maybe ExprType`.
A good name for that might be `getValue`.
Unfortunately, `getValue` is taken.
Instead of stymieing on naming, we write this and push off to the future.

joneshf-cn committed:
PS-3032 Inline `View` in `Input`

We're about to make some changes to what `Input` is.
To make it clear what is really changing,
we inline `View` so we're not muddying this expansion
in a useful diff.

joneshf-cn committed:
PS-3032 Add rows to display for the list view

We've got all the pieces now.
We can take a collection of `Page _`s,
and display them as rows in our list view.

This is mostly straight forward, we had to thread through some values.

davezuch commented on src/Lynx/List.purs:
```diff
@@ -67,14 +69,17 @@ type Column
 
 type Input
   = { columns :: Array Column
+    , rows :: Array (Lynx.Form.Page Lynx.Expr.ExprType)
```
Can we assume that we're getting `Page ExprType`? I think the `GET` endpoint will return `Page Expr`, since that's what the form will expect, and may even be what we'll expect here. I guess this is one of those things we can kick down the road, but I do think we'll end up evaluating each row before rendering it.

joneshf-cn replied:
Hmm, I'm not sure. This is what I was thinking: we can only create orders when they have all the values, those values would have to be someone typing in strings that we're evaluating, so we'd be storing the values as `ExprType` not parsing them to `Expr`. Is that not how it's going to work?

davezuch replied:
> we can only create orders when they have all the values

Only when we have `required :: Boolean true` on all fields. But Spencer doesn't want us to worry about validation for now, so we're saving everything. Also, if every field has a `default :: Just _`, then we have values anyway. 

> so we'd be storing the values as `ExprType` not parsing them to `Expr`. Is that not how it's going to work?

I think we should always store values as `ExprType`, even for `Page Expr`. I don't think a value of `InputSource Expr` ever makes sense, as users can't input expressions (maybe in the Form Builder form it'd make sense). 

But returning `Page Expr` instead of `Page ExprType` is necessary, at least for the form, because it needs to know what the expressions are in order to evaluate changes to the form. At the least we'd need to send both, but that seems redundant because we can get one from the other.

For the list view it's less clear that `Page Expr` is necessary, but I argue that it is, because evaluation isn't pure, there are HTTP lookups that can happen, or date validations that might happen, which means the result of evaluating now might be different from the result of evaluating when the Order was saved. I guess which result is more correct is debatable, but I think the newer one.

joneshf-cn replied:
I see. It's important to note that whole thing isn't accurate anyway. We should be working with an Order, not a `Page _` (whatever that is in the OMS parlance). I think no matter what we put in here, it'll be an approximation of what the real thing is until we get to that.

davezuch reviewed (approved):
Changes look good!
